packages/babel-helper-modules/src/normalize-and-load-metadata.js

```diff
@@ -0,0 +1,378 @@
+          if (reexport) {
+            localData.delete(localName);
+
+            reexport.names.forEach(name => {
+              data.reexports.set(name, "default");
+            });
+          }
+        } else if (spec.isImportNamespaceSpecifier()) {
+          const localName = spec.get("local").node.name;
+
+          if (!strict) data.interop = "namespace";
+
+          data.importsNamespace.add(localName);
+          const reexport = localData.get(localName);
+          if (reexport) {
+            localData.delete(localName);
+
+            reexport.names.forEach(name => {
+              data.reexportNamespace.add(name);
+            });
+          }
+        } else if (spec.isImportSpecifier()) {
+          const importName = spec.get("imported").node.name;
+          const localName = spec.get("local").node.name;
+
+          data.imports.set(localName, importName);
+
+          if (importName === "default" && data.interop === "none") {
+            data.interop = "default";
+          }
+
+          const reexport = localData.get(localName);
+          if (reexport) {
+            localData.delete(localName);
+
+            reexport.names.forEach(name => {
+              data.reexports.set(name, importName);
+            });
+          }
+        }
+      });
+      child.remove();
+    } else if (child.isExportAllDeclaration()) {
+      const data = getData(child.node.source);
+      data.statements.push(child.node);
+
+      data.reexportAll = {
+        loc: child.node.loc,
+      };
+      child.remove();
+    } else if (child.isExportNamedDeclaration()) {
+      const data = getData(child.node.source);
+      data.statements.push(child.node);
+
+      child.get("specifiers").forEach(spec => {
+        if (!spec.isExportSpecifier()) {
+          throw spec.buildCodeFrameError("Unexpected export specifier type");
+        }
+        const importName = spec.get("local").node.name;
+        const exportName = spec.get("exported").node.name;
+
+        if (importName === "default" && data.interop === "none") {
+          data.interop = "default";
+        }
+
+        data.reexports.set(exportName, importName);
+
+        if (exportName === "__esModule") {
+          throw exportName.buildCodeFrameError('Illegal export "__esModule".');
+        }
+      });
+      child.remove();
+    }
+  });
+
+  // Reuse the imported namespace name if there is one.
+  for (const [, metadata] of sourceData) {
+    if (metadata.importsNamespace.size > 0) {
+      [metadata.name] = metadata.importsNamespace;
+    }
+
+    if (noInterop) metadata.interop = "none";
+  }
+
+  return {
+    exportName,
+    local: localData,
+    source: sourceData,
+  };
+}
+
+/**
+ * Remove all local exports, leaving only re-export statements like these:
+ *  * "export {foo} from 'source';"
+ *  * "export * from 'source';"
+ */
+function normalizeLocalExports(
+  programPath: NodePath,
+): Map<string, LocalExportMetadata> {
```

Comment:
Yeah I was on the fence. Since the default export of this module _does_ return metadata and mutate, I didn't worry about it too much since we won't be making a breaking change if we come refactor it later. That said, I'm happy to change that now if you want.

Reply:
Ideally we'd have a method to collect metadata without mutations, and a method to apply the mutations as directed by the metadata, for each of the cases; but I understand if they have to be kept "dirty" due to the performance impact of multiple traversals.